tensorstore/kvstore/s3/s3_key_value_store.cc

```diff
@@ -0,0 +1,1306 @@
+#include "tensorstore/kvstore/gcs/validate.h"
+#include "tensorstore/kvstore/s3/s3_resource.h"
+#include "tensorstore/kvstore/s3/s3_request_builder.h"
+#include "tensorstore/kvstore/s3/s3_credential_provider.h"
+#include "tensorstore/kvstore/s3/s3_metadata.h"
+#include "tensorstore/kvstore/s3/validate.h"
+#include "tensorstore/kvstore/s3/s3_uri_utils.h"
+#include "tensorstore/util/execution/any_receiver.h"
+#include "tensorstore/util/execution/execution.h"
+#include "tensorstore/util/executor.h"
+#include "tensorstore/util/future.h"
+#include "tensorstore/util/garbage_collection/fwd.h"
+#include "tensorstore/util/quote_string.h"
+#include "tensorstore/util/result.h"
+#include "tensorstore/util/status.h"
+#include "tensorstore/util/str_cat.h"
+
+
+// specializations
+#include "tensorstore/internal/cache_key/std_optional.h"
+#include "tensorstore/internal/json_binding/std_array.h"
+#include "tensorstore/internal/json_binding/std_optional.h"
+#include "tensorstore/util/garbage_collection/std_optional.h"
+#include "tensorstore/serialization/fwd.h"
+#include "tensorstore/serialization/std_optional.h"
+
+
+using ::tensorstore::internal::IntrusivePtr;
+using ::tensorstore::internal::DataCopyConcurrencyResource;
+using ::tensorstore::internal::ScheduleAt;
+using ::tensorstore::internal::SHA256Digester;
+using ::tensorstore::internal::MD5Digester;
+using ::tensorstore::internal_http::HttpRequest;
+using ::tensorstore::internal_http::HttpResponse;
+using ::tensorstore::internal_http::HttpTransport;
+using ::tensorstore::internal_auth_s3::S3Credentials;
+using ::tensorstore::internal_auth_s3::CredentialProvider;
+using ::tensorstore::internal_auth_s3::GetS3CredentialProvider;
+using ::tensorstore::internal_storage_s3::StorageGenerationFromHeaders;
+using ::tensorstore::internal_storage_gcs::IsRetriable;
+using ::tensorstore::internal_kvstore_gcs_http::RateLimiter;
+using ::tensorstore::internal_kvstore_gcs_http::RateLimiterNode;
+using ::tensorstore::internal_storage_s3::S3ConcurrencyResource;
+using ::tensorstore::internal_storage_s3::S3RateLimiterResource;
+using ::tensorstore::internal_storage_s3::S3RequestRetries;
+using ::tensorstore::internal_storage_s3::S3RequestBuilder;
+using ::tensorstore::internal_storage_s3::IsValidBucketName;
+using ::tensorstore::internal_storage_s3::IsValidObjectName;
+using ::tensorstore::internal_storage_s3::IsValidStorageGeneration;
+using ::tensorstore::internal_storage_s3::GetTag;
+using ::tensorstore::internal_storage_s3::FindTag;
+using ::tensorstore::internal_http_s3::S3UriEncode;
+using ::tensorstore::internal_http_s3::S3UriObjectKeyEncode;
+using ::tensorstore::kvstore::Key;
+using ::tensorstore::kvstore::ListOptions;
+using ::tensorstore::kvstore::SupportedFeatures;
+
+
+#ifndef TENSORSTORE_INTERNAL_S3_LOG_REQUESTS
+#define TENSORSTORE_INTERNAL_S3_LOG_REQUESTS 0
+#endif
+
+#ifndef TENSORSTORE_INTERNAL_S3_LOG_RESPONSES
+#define TENSORSTORE_INTERNAL_S3_LOG_RESPONSES 0
+#endif
+
+
+namespace {
```

Comment:
You can put these inside the inner nested anonymous namespace.

Reply:
done